mlir/lib/Conversion/ConvertToAIRPass.cpp

```diff
@@ -2318,6 +2318,267 @@ class ScfForallToLaunchConversion : public OpRewritePattern<scf::ForallOp> {
+        src_memref_ty =
+            memref::SubViewOp::inferRankReducedResultType(
+                subviewOp.getType().getShape(), src_memref_ty,
+                subviewOp.getStaticOffsets(), subviewOp.getStaticSizes(),
+                subviewOp.getStaticStrides())
+                .cast<MemRefType>();
+      else
+        src_memref_ty =
+            memref::SubViewOp::inferResultType(
+                src_memref_ty, subviewOp.getStaticOffsets(),
+                subviewOp.getStaticSizes(), subviewOp.getStaticStrides())
+                .cast<MemRefType>();
+    }
+  }
+
+  for (auto memrefOp : dst_ancestor_memref_ops) {
+    if (auto transposeOp = dyn_cast<memref::TransposeOp>(memrefOp)) {
+      dst_memref_ty =
+          inferTransposeResultType(dst_memref_ty, transposeOp.getPermutation());
+    } else if (auto expandShapeOp = dyn_cast<memref::ExpandShapeOp>(memrefOp)) {
+      FailureOr<MemRefType> compute_expand =
+          memref::ExpandShapeOp::computeExpandedType(
+              dst_memref_ty, expandShapeOp.getResultType().getShape(),
+              expandShapeOp.getReassociationIndices());
+      if (failed(compute_expand)) {
+        assert(false);
+      } else {
+        dst_memref_ty = *compute_expand;
+      }
+    } else if (auto subviewOp = dyn_cast<memref::SubViewOp>(memrefOp)) {
+      if (subviewOp.getSourceType().getRank() > subviewOp.getType().getRank())
+        dst_memref_ty =
+            memref::SubViewOp::inferRankReducedResultType(
+                subviewOp.getType().getShape(), dst_memref_ty,
+                subviewOp.getStaticOffsets(), subviewOp.getStaticSizes(),
+                subviewOp.getStaticStrides())
+                .cast<MemRefType>();
+      else
+        dst_memref_ty =
+            memref::SubViewOp::inferResultType(
+                dst_memref_ty, subviewOp.getStaticOffsets(),
+                subviewOp.getStaticSizes(), subviewOp.getStaticStrides())
+                .cast<MemRefType>();
+    }
+  }
+
+  if (src_ancestor_memref_ops.size()) {
+    extractStridesFromMemrefType(src_memref_ty, rewriter, src_strides);
+    extractSizesFromMemrefType(src_memref_ty, rewriter, src_sizes);
+  }
+  if (dst_ancestor_memref_ops.size()) {
+    extractStridesFromMemrefType(dst_memref_ty, rewriter, dst_strides);
+    extractSizesFromMemrefType(dst_memref_ty, rewriter, dst_sizes);
+  }
+
+  SmallVector<Value, 4> deps;
+  SmallVector<Type, 4> tys;
+
+  if (failed(canonicalizeAIRDmaOperands(rewriter, src_offsets, src_sizes,
+                                        src_strides,
+                                        src.getType().cast<MemRefType>())) ||
+      failed(canonicalizeAIRDmaOperands(rewriter, dst_offsets, dst_sizes,
+                                        dst_strides,
+                                        dst.getType().cast<MemRefType>()))) {
+    assert(false);
+  }
+  auto new_dma = rewriter.create<xilinx::air::DmaMemcpyNdOp>(
```

Comment:
Can you please explain what the new source and destination tensors are? 

I would have expected them to be different to the old DmaMemcpyNdOp (as you've unwound through view operations). 

i.e. it seems to me like you've updated the DmMemcpy attributes to reflect the preceding view operations, but the view operations are still being applied (so effectively applied twice now). I'm probably missing something

Reply:
The original memcpy op has src and dst memrefs being those returned at the end of chains of memref ops. The src and dst memrefs of the new memcpy op have been changed to the starting memrefs of each chain, where the chain of memref ops are condensed into sizes and strides list and applied at the two sides of the new memcpy. The chain of memref ops are then removed.

Reply:
Ok good. I can see that now (I somehow missed the use of dst_ancestor_memref_ops[0] and src_ancestor_memref_ops[0] when I looked earlier).